miio/viomivacuum.py

```diff
@@ -268,124 +380,398 @@ def has_map(self) -> bool:
+    @command(
+        click.option(
+            "--rooms",
+            "-r",
+            multiple=True,
+            help="Rooms name or room id. Can be used multiple times",
+        )
+    )
+    def start_with_room(self, rooms):
+        """Start cleaning specific rooms."""
+        if not self._cache["rooms"]:
+            self.get_rooms()
+        reverse_rooms = {v: k for k, v in self._cache["rooms"].items()}
+        room_ids = []
+        for room in rooms:
+            if room in self._cache["rooms"]:
+                room_ids.append(room)
+            elif room in reverse_rooms:
+                room_ids.append(reverse_rooms[room])
+            else:
+                return "Rooms {} is unknown, it should be in '{}' " "or in '{}'".format(
+                    room,
+                    ", ".join(self._cache["rooms"].keys()),
+                    ", ".join(self._cache["rooms"].values()),
+                )
+        self._cache["edge_state"] = self.get_properties(["mode"])
+        self.send(
+            "set_mode_withroom",
+            self._cache["edge_state"] + [1, 0, len(room_ids)] + room_ids,
+        )
 
     @command()
     def pause(self):
         """Pause cleaning."""
-        self.send("set_mode_withroom", [0, 2, 0])
+        # params: [edge_state, 0]
+        # - edge: see ViomiEdgeState
+        # - 2: pause cleaning
+        if not self._cache["edge_state"]:
+            self._cache["edge_state"] = self.get_properties(["mode"])
+        self.send("set_mode", self._cache["edge_state"] + [2])
+
+    @command()
+    def stop(self):
+        """Validate that Stop cleaning."""
+        # params: [edge_state, 0]
+        # - edge: see ViomiEdgeState
+        # - 0: stop cleaning
+        if not self._cache["edge_state"]:
+            self._cache["edge_state"] = self.get_properties(["mode"])
+        self.send("set_mode", self._cache["edge_state"] + [0])
+
+    @command(click.argument("mode", type=EnumType(ViomiMode)))
+    def clean_mode(self, mode: ViomiMode):
+        """Set the cleaning mode.
+
+        [vacuum, vacuumAndMop, mop, cleanzone, cleanspot]
+        """
+        self.send("set_mop", [mode.value])
 
     @command(click.argument("speed", type=EnumType(ViomiVacuumSpeed)))
     def set_fan_speed(self, speed: ViomiVacuumSpeed):
         """Set fanspeed [silent, standard, medium, turbo]."""
         self.send("set_suction", [speed.value])
 
-    @command(click.argument("watergrade"))
+    @command(click.argument("watergrade", type=EnumType(ViomiWaterGrade)))
     def set_water_grade(self, watergrade: ViomiWaterGrade):
-        """Set water grade [low, medium, high]."""
+        """Set water grade.
+
+        [low, medium, high]
+        """
         self.send("set_suction", [watergrade.value])
 
+    def get_positions(self, plan_multiplicator=1) -> [ViomiPositionPoint]:
+        """Return the last positions.
+
+        returns: [x, y, phi, update, x, y, phi, update, x, y, phi, update, ...]
```

Comment:
```suggestion
```
This can be removed, as the type hint already reveals the return type.

The plan_multiplicator should be documented, though.

Reply:
done